prov/gni/src/gnix_dom.c

```diff
@@ -89,11 +89,26 @@ static int gnix_domain_close(fid_t fid)
 		}
 	}
 
+	/*
+	 *  remove nics from the domain's nic list,
+	 *  decrement ref_cnt on each nic.  If ref_cnt
+	 *  drops to 0, destroy the cdm, remove from
+	 *  the global nic list.
+	 */
 	list_for_each_safe(&domain->nic_list, p, next, list)
 	{
 		list_del(&p->list);
 		gnix_list_node_init(&p->list);
-		/* TODO: free nic here */
+		atomic_dec(&p->ref_cnt);
+		assert(atomic_get(&p->ref_cnt) >= 0);
+		if (atomic_get(&p->ref_cnt) == 0) {
```

Comment:
I think this risks a double/multiple free situation.  It would be better to capture the result from the `atomic_dec` call above and test that value.  That same value can be used for the assertion.